lib/Differentiator/StmtClone.cpp

```diff
@@ -520,8 +520,12 @@ bool ReferencesUpdater::VisitDeclRefExpr(DeclRefExpr* DRE) {
   // Replace the declaration if it is present in `m_DeclReplacements`.
   if (VarDecl* VD = dyn_cast<VarDecl>(DRE->getDecl())) {
     auto it = m_DeclReplacements.find(VD);
-    if (it != std::end(m_DeclReplacements))
+    if (it != std::end(m_DeclReplacements)) {
       DRE->setDecl(it->second);
+      if (it->second->getType().getNonReferenceType()!=DRE->getType()) {
+        DRE->setType(it->second->getType().getNonReferenceType());
+      }
+    }
```

Comment:
```suggestion
      QualType NonRefQT = it->second->getType().getNonReferenceType();
      if (NonRefQT != DRE->getType())
        DRE->setType(NonRefQT);
    }
```

Why this change was necessary? Can it go in a separate PR with a test?

Reply:
To make ref-type declarations global, I replaced them with pointers (to allow us to separate the declaration and the initialization). However, ``VisitorBase::Clone`` didn't expect a cloned expression to have a different type than the original expression, which led to broken AST. This could go in a separate PR but it doesn't make much sense by itself.

Reply:
Ok, the code suggestion still makes sense though..